scipy/stats/tests/test_stats.py

```diff
@@ -5668,3 +5673,133 @@ def test_dist_perm(self):
+          (5, 10, 1703, 99.9, lt, 'asymptotic', [array([10, 9, 8, 7, 6, 5, 4, 3, 2, 1]), array([10, 9, 8, 7, 6, 5, 4, 3, 2, 1]), array([10, 9, 8, 7, 6, 5, 4, 3, 2, 1]), array([9, 2, 8, 7, 6, 5, 4, 3, 10, 1]), array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])]),
+          (5, 10, 1702, 99.9, gt, 'asymptotic', [array([10, 9, 8, 7, 6, 5, 4, 3, 2, 1]), array([10, 9, 8, 7, 6, 5, 4, 3, 2, 1]), array([10, 9, 8, 7, 6, 5, 4, 3, 2, 1]), array([9, 2, 7, 8, 6, 5, 4, 3, 10, 1]), array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])]),
+          (3, 6, 244, 95, lt, 'exact', [array([6, 5, 4, 3, 2, 1]), array([6, 5, 4, 3, 2, 1]), array([1, 3, 4, 5, 2, 6])]),
+          (3, 6, 243, 95, gt, 'exact', [array([6, 5, 4, 3, 2, 1]), array([6, 5, 4, 3, 2, 1]), array([1, 4, 2, 5, 3, 6])]),
+          (10, 5, 477, 95, lt, 'exact', [array([5, 4, 3, 2, 1]), array([5, 4, 3, 2, 1]), array([5, 4, 3, 2, 1]), array([5, 4, 3, 2, 1]), array([5, 4, 3, 2, 1]), array([5, 4, 3, 2, 1]), array([4, 1, 3, 2, 5]), array([1, 2, 3, 4, 5]), array([1, 2, 3, 4, 5]), array([1, 2, 3, 4, 5])]),
+          (10, 5, 476, 95, gt, 'exact', [array([5, 4, 3, 2, 1]), array([5, 4, 3, 2, 1]), array([5, 4, 3, 2, 1]), array([5, 4, 3, 2, 1]), array([5, 4, 3, 2, 1]), array([5, 4, 3, 2, 1]), array([4, 1, 2, 3, 5]), array([1, 2, 3, 4, 5]), array([1, 2, 3, 4, 5]), array([1, 2, 3, 4, 5])]),
+          (19, 3, 239, 95, lt, 'exact', [array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([2, 1, 3]), array([1, 2, 3]), array([1, 2, 3]), array([1, 2, 3]), array([1, 2, 3]), array([1, 2, 3]), array([1, 2, 3])]),
+          (19, 3, 238, 95, gt, 'exact', [array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([1, 2, 3]), array([1, 2, 3]), array([1, 2, 3]), array([1, 2, 3]), array([1, 2, 3]), array([1, 2, 3]), array([1, 2, 3])]),
+          (4, 7, 501, 99, lt, 'exact', [array([7, 6, 5, 4, 3, 2, 1]), array([7, 6, 5, 4, 3, 2, 1]), array([6, 5, 7, 4, 3, 2, 1]), array([1, 2, 3, 4, 5, 6, 7])]),
+          (4, 7, 500, 99, gt, 'exact', [array([7, 6, 5, 4, 3, 2, 1]), array([7, 6, 5, 4, 3, 2, 1]), array([5, 6, 7, 4, 3, 2, 1]), array([1, 2, 3, 4, 5, 6, 7])]),
+          (11, 3, 147, 99.9, lt, 'exact', [array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([2, 1, 3]), array([1, 2, 3])]),
+          (11, 3, 146, 99.9, gt, 'exact', [array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([3, 2, 1]), array([1, 2, 3]), array([1, 2, 3])]),
+         ]
```

Comment:
I originally the tests above when I was using a Monte Carlo method instead of "exact". It was a reasonable way of testing considering that Monte Carlo results may vary.

When I implemented the exact method, I found that reference [5] has exact p values listed for some values of L, so I could compare against those for the exact method.

I could compare the asymptotic p-values against those from some other software.

Would that be preferable?

Reply:
Sure, that would be great. I imagine you've already written the code, so if you'd like, feel free to push the tests to this branch directly or make a PR on my repo.

Reply:
I don't have a complete test suite based on it, but here's R code that can be used to generate test cases.  It works in [`rdrr.io`](https://rdrr.io/snippets/).  `page.test` uses the exact distribution for k up to 22.  This example has 25 columns, so the approximate p-value is computed:
```
library(cultevo)

data = rbind(c(72, 47, 73, 35, 47, 96, 30, 59, 41, 36, 56, 49, 81, 43, 70, 47,
               28, 28, 62, 20, 61, 20, 80, 24, 50),
             c(68, 52, 60, 34, 44, 20, 65, 88, 21, 81, 48, 31, 31, 67, 69, 94,
               30, 24, 40, 87, 70, 43, 50, 96, 43),
             c(81, 13, 85, 35, 79, 12, 92, 86, 21, 64, 16, 64, 68, 17, 16, 89,
               71, 43, 43, 36, 54, 13, 66, 51, 55))

result = page.test(data, verbose=FALSE)
result$statistic
result$p.value
result$p.type
```
Output
```
Exact p value is not available for k=25, using Normal approximation.
    L 
12332 
[1] 0.7722477
[1] "approximate"
```
When I change `data` to
```
data = rbind(c(1.5,  4. ,  8.3,  5. , 19. , 11.),
             c(5. ,  4. ,  3.5, 10. , 20. , 21.),
             c(8.4,  3.2, 10. , 12. , 14. , 15.))
```
the output is
```
  L 
266 
[1] 4.121656e-05
[1] "exact"
```

Reply:
Great, thanks.